spring-core/src/main/java/org/springframework/core/io/support/PathMatchingResourcePatternResolver.java

```diff
@@ -712,31 +711,58 @@ protected JarFile getJarFile(String jarFileUrl) throws IOException {
 			}
 			return Collections.emptySet();
 		}
-		return doFindMatchingFileSystemResources(rootDir, subPattern);
+		return doFindMatchingFileSystemResources(rootDirResource, rootDir, subPattern);
 	}
 
 	/**
 	 * Find all resources in the file system that match the given location pattern
 	 * via the Ant-style PathMatcher.
+	 * @param rootDirResource the root directory as Resource
 	 * @param rootDir the root directory in the file system
 	 * @param subPattern the sub pattern to match (below the root directory)
 	 * @return a mutable Set of matching Resource instances
 	 * @throws IOException in case of I/O errors
 	 * @see #retrieveMatchingFiles
 	 * @see org.springframework.util.PathMatcher
 	 */
-	protected Set<Resource> doFindMatchingFileSystemResources(File rootDir, String subPattern) throws IOException {
+	protected Set<Resource> doFindMatchingFileSystemResources(Resource rootDirResource, File rootDir, String subPattern) throws IOException {
 		if (logger.isTraceEnabled()) {
 			logger.trace("Looking for matching resources in directory tree [" + rootDir.getPath() + "]");
 		}
 		Set<File> matchingFiles = retrieveMatchingFiles(rootDir, subPattern);
 		Set<Resource> result = new LinkedHashSet<>(matchingFiles.size());
 		for (File file : matchingFiles) {
-			result.add(new FileSystemResource(file));
+			String relativePath = relativeTo(file, rootDir);
+			result.add(rootDirResource.createRelative(relativePath));
 		}
 		return result;
 	}
 
+	/**
+	 * Compute the relative path for given file based on the given root directory.
+	 * @param file the file to get relative path
+	 * @param rootDir the root file
+	 * @return the relative path
+	 */
+	private String relativeTo(File file, File rootDir) {
+		String filePath = getAbsolutePath(file);
+		String rootPath = getAbsolutePath(rootDir);
+		return filePath.substring(rootPath.length());
+	}
+
+	/**
+	 * Get the given file's absolute path. If the file is a directory then a slash is appended.
+	 * @param file the file to get absolute path
+	 * @return the absolute path
+	 */
+	private String getAbsolutePath(File file) {
+		String absolutePath = StringUtils.replace(file.getAbsolutePath(), File.separator, "/");
```

Comment:
Why do you want to replace the `File.separator`, which is system specific, with a `"/"`? Even though it _should_ be safe to use because it _should_ be translated by any libraries to a platform-specific file separator.